### PR TITLE
NDRS-672: Track validator weights when syncing

### PR DESCRIPTION
**Pull request by xcthulhu**

### CODE REVIEW

**Comment by xcthulhu** [comment]
Tracking validator weights when syncing is close to done, but there's a few dangling things:

 1. @henrytill @EdHastingsCasperLabs I need to change the protocol buffer for `StepResult`, but I am not sure what that entails.
 2. @goral09 @fizyk20 @afck When a blockheader comes over the wire, we should reject it if it has the wrong era validators.  I'm not sure how to do this and maybe this is another ticket.

File: execution_engine/src/core/engine_state/mod.rs
Comment (xcthulhu) [inline execution_engine/src/core/engine_state/mod.rs:2479]: @EdHastingsCasperLabs It ended up being easiest to have this become a `GetEraValidatorsError`, but it's a bit odd and I'm not sure if there's a better way here.

File: grpc/server/src/engine_server/mod.rs
Comment (xcthulhu) [inline grpc/server/src/engine_server/mod.rs:455]: @henrytill @EdHastingsCasperLabs including `next_era_validators` in the protocol buffers drastically changes their format.  Not including them doesn't seem to break any tests... what should I expect to break here?

File: node/src/components/block_executor.rs
Comment (xcthulhu) [inline node/src/components/block_executor.rs:553]: @afck @fizyk20 @goral09 Feels bad to do an `expect` here... is there something better I can do?

Comment (xcthulhu) [reply node/src/components/block_executor.rs:553]: Perhaps a `filter_map` would be better here...?

File: grpc/server/src/engine_server/mod.rs
Comment (xcthulhu) [reply grpc/server/src/engine_server/mod.rs]: I've updated the protocol buffers in https://github.com/CasperLabs/casper-node/pull/717/commits/67140cb59f3dcf950bb8a6402196ee6d72f03341

**Comment by xcthulhu** [comment]
>     1. @henrytill @EdHastingsCasperLabs I need to change the protocol buffer for `StepResult`, but I am not sure what that entails.

To answer this question, the protocol buffers are used in the tests.  I have updated the protocol buffers here: https://github.com/CasperLabs/casper-node/pull/717/commits/67140cb59f3dcf950bb8a6402196ee6d72f03341

File: node/src/components/block_executor.rs
Comment (afck) [reply node/src/components/block_executor.rs:553]: That can only fail if a validator bonds with an invalid public key, can it?
I think we have the same problem in `EraSupervisor` somewhere; we just need to make sure we do the same thing in both places.

(Also: formatting! :grin:)

Comment (xcthulhu) [reply node/src/components/block_executor.rs]: Fixed in https://github.com/CasperLabs/casper-node/pull/717/commits/86eff05cf00063166109268eaaaf6e1f5436cc06

File: execution_engine/src/core/engine_state/mod.rs
Comment (xcthulhu) [reply execution_engine/src/core/engine_state/mod.rs:2550]: `into_bytes` gives a `Result<_, bytesrepr::Error>`.  But `GetEraValidatorsError` doesn't have a variant for `bytesrepr::Error`.  There is a variant for `engine_state::error::Error`, and `engine_state::error::Error` has a variant for `bytesrepr::Error`, so we do two coercions.

I could modify `GetEraValidatorsError` but it was simpler to do this.

File: execution_engine/src/core/engine_state/mod.rs
Comment (xcthulhu) [reply execution_engine/src/core/engine_state/mod.rs:2686]: Again, I did this to avoid extending the `GetEraValidators` error enum.

File: grpc/server/src/engine_server/mod.rs
Comment (xcthulhu) [reply grpc/server/src/engine_server/mod.rs:469]: So in this leg here:

```rust
            Ok(result) => {
                let err_msg = result.to_string();
                warn!("{}", err_msg);
                let mut ret = ipc::StepResponse::new();
                ret.mut_step_result().mut_error().set_message(err_msg);
                return SingleResponse::completed(ret);
            }
```

Here `result` is a `StepResult`. But it is not `StepResult::Success` which is handled as the first `match` case.  When we convert to protobuf we make it an error.

The other path for dealing with `GetErrorValidatorsErrors` has the same logic but the types are different.  Not sure how to dry this code out exactly.

**Review by afck — APPROVED** [review]
Looks good to me, but let's also wait for someone more familiar with the Contract Runtime to review it.

**Comment by henrytill** [comment]
As discussed offline, you should restore the broken step tests so that @EdHastingsCasperLabs can take a look.

**Review by EdHastingsCasperLabs — COMMENTED** [review]
I'm digging into this now; should have a resolution soon.

**Comment by xcthulhu** [comment]
Thanks @EdHastingsCasperLabs for the fixup... I want to wait for your approval before going ahead with the merge.

File: node/src/types/block.rs
Comment (goral09) [inline node/src/types/block.rs:561]: Why is it an `Option`? We should be able to always populate it.

**Comment by goral09** [comment]
> 2. @goral09 @fizyk20 @afck When a blockheader comes over the wire, we should reject it if it has the wrong era validators.  I'm not sure how to do this and maybe this is another ticket.

We don't send linear chain blocks over the wire so we don't have to validate them. We only create finality signatures for them, which are gossiped, and they are validated.

File: node/src/components/linear_chain_sync.rs
Comment (goral09) [inline node/src/components/linear_chain_sync.rs:305]: I think the approach with conditions on match patterns is dangerous as it's error-prone. If this and the next arms were switched places then the logic changes. That's why I put it under one match case. It was deliberate.

**Pull request merged**
